File: src/BaseHmiTypes/Converters/AutomationXml/HmiCycleToAutomationXmlConverter.cs
using System.Globalization;
using System.Xml;
using BaseHmiTypes.Cycles;

namespace BaseHmiTypes.Converters.AutomationXml;

public static class HmiCycleToAutomationXmlConverter
{
    public static void WriteAutomationXml(XmlWriter writer, HmiCycle cycle, ref int id)
    {
        writer.WriteStartElement("Hmi.Cycle.Cycle");
        writer.WriteAttributeString("ID", (id++).ToString(CultureInfo.InvariantCulture));
        writer.WriteStartElement("AttributeList");
        WriteElementString(writer, "CycleTime", cycle.CycleTime);
        WriteElementString(writer, "CycleUnit", cycle.CycleUnit);
        WriteElementString(writer, "Name", cycle.Name);
        WriteElementString(writer, "StartAtStartingPoint", cycle.StartAtStartingPoint);
        WriteElementString(writer, "StartingPoint", "****-" + WriteEnumValue(cycle.StartingPointMonth) + "-" + WriteEnumValue(cycle.StartingPointDay) + "T" + WriteEnumValue(cycle.StartingPointHour) + ":" + WriteEnumValue(cycle.StartingPointMinute) + ":" + WriteEnumValue(cycle.StartingPointSecond));
        WriteElementString(writer, "TriggerAtShutdown", cycle.TriggerAtShutDown);
        WriteElementString(writer, "TriggerAtStartup", cycle.TriggerAtStartUp);
        writer.WriteEndElement();

        if (cycle.Comment != null)
        {
            writer.WriteStartElement("ObjectList");
            HmiAutomationXmlConverter.WriteAutomationXml(cycle.Comment, writer, ref id, "Comment");
            writer.WriteEndElement();
        }

        writer.WriteEndElement();
    }

    private static string WriteEnumValue(int value)
    {
        return value > 999 ? "**" : value.ToString(CultureInfo.InvariantCulture).PadLeft(2, '*');
    }

    private static void WriteElementString(XmlWriter writer, string name, object? value)
    {
        writer.WriteElementString(name, value switch
        {
            null => string.Empty,
            bool boolean => boolean ? "true" : "false",
            IFormattable formattable => formattable.ToString(null, CultureInfo.InvariantCulture),
            _ => value.ToString()
        });
    }
}

File: typescript/base-hmi-types/src/converters/automation-xml/HmiCycleToAutomationXmlConverter.ts
import { HmiCycle } from "../../cycles/HmiCycle.js";
import { HmiAutomationXmlConverter } from "./HmiAutomationXmlConverter.js";
import { XmlWriter } from "./XmlWriter.js";

export class HmiCycleToAutomationXmlConverter {
  static writeAutomationXml(writer: XmlWriter, cycle: HmiCycle, id: { value: number }): void {
    writer.writeStartElement("Hmi.Cycle.Cycle");
    writer.writeAttributeString("ID", String(id.value++));
    writer.writeStartElement("AttributeList");
    writer.writeElementString("CycleTime", cycle.cycleTime);
    writer.writeElementString("CycleUnit", cycle.cycleUnit);
    writer.writeElementString("Name", cycle.name);
    writer.writeElementString("StartAtStartingPoint", cycle.startAtStartingPoint);
    writer.writeElementString("StartingPoint", `****-${writeEnumValue(cycle.startingPointMonth)}-${writeEnumValue(cycle.startingPointDay)}T${writeEnumValue(cycle.startingPointHour)}:${writeEnumValue(cycle.startingPointMinute)}:${writeEnumValue(cycle.startingPointSecond)}`);
    writer.writeElementString("TriggerAtShutdown", cycle.triggerAtShutDown);
    writer.writeElementString("TriggerAtStartup", cycle.triggerAtStartUp);
    writer.writeEndElement("AttributeList");

    if (cycle.comment !== undefined) {
      writer.writeStartElement("ObjectList");
      HmiAutomationXmlConverter.writeMultilingualText(cycle.comment, writer, id, "Comment");
      writer.writeEndElement("ObjectList");
    }

    writer.writeEndElement("Hmi.Cycle.Cycle");
  }
}

function writeEnumValue(value: number): string {
  return value > 999 ? "**" : String(value).padStart(2, "*");
}
